### ISSUE TITLE
RFC: should ++ be generalized to mean concatenation of iterators?

### ISSUE THREAD

**Issue by simeonschaub** [tlienart/PlusPlus.jl#3]
Just a random thought: I think it might be nice to formalize what `++` should mean, and I think the gist of it should be that it concatenates iterators, using more appropriate types to represent the result if applicable. To be consistent, this would have to mean concatenating arrays along their last dimension instead of the first, since that will give the right iteration order. This seems more useful to me for a column-major layout anyways. Here's a rough sketch of what I'm thinking of:
```julia
struct Concat{T<:Tuple}
    iters::T
end
function Base.iterate(c::Concat, st=(1,))
    i, st_i = st[1], Base.tail(st)
    i > length(c.iters) && return
    x = iterate(c.iters[i], st_i...)
    x === nothing && return iterate(c, (i+1,))
    return x[1], (i, x[2])
end
Base.IteratorSize(::Concat) = Base.SizeUnknown()
++() = Concat(())
++(x...) = Concat(x)
++(x::Union{AbstractArray, Number}...) = cat(x...; dims=max(1, maximum(ndims, x)))
++(x::Union{AbstractString, AbstractChar}...) = string(x...)
++(x::Tuple...) = foldl((x, y) -> (x..., y...), x)
```
```julia
julia> 1 ++ [3 4] ++ [5]
1×4 Array{Int64,2}:
 1  3  4  5

julia> "a" ++ 'b'
"ab"

julia> "a" ++ (x^2 for x in 1:3) ++ (:foo,)
Concat{Tuple{String,Base.Generator{UnitRange{Int64},var"#7#8"},Tuple{Symbol}}}(("a", Base.Generator{UnitRange{Int64},var"#7#8"}(var"#7#8"(), 1:3), (:foo,)))

julia> collect(ans)
5-element Array{Any,1}:
  'a': ASCII/Unicode U+0061 (category Ll: Letter, lowercase)
 1
 4
 9
  :foo
```
What do you think about that idea? Could of course also just be a separate package, if you think that's better.

**Comment by CameronBieganek** [comment]
Good idea! Your example shows that the current PlusPlus implementation is a little lacking in some regards:

```julia
julia> using PlusPlus

julia> "a" ++ (x^2 for x in 1:3) ++ (:foo,)
3-element Array{Any,1}:
 "a"
 Base.Generator{UnitRange{Int64},var"#1#2"}(var"#1#2"(), 1:3)
 (:foo,)
```

So, right now `PlusPlus` just packs those three iterators into an array. 🤣 

It appears that I suggested something related [in this comment on Discourse](https://discourse.julialang.org/t/type-inference-for-nested-iterators/47082/17?u=cameronbieganek). I used type promotion in that version. Also, [this Github issue](https://github.com/JuliaLang/julia/issues/36760#issuecomment-698698602) is related.

**Comment by tlienart** [comment]
Thanks for the suggestion! Generalising and formalising seems useful :)
Maybe open a PR? 

I'll try to play with your suggested implementation today to get familiar with it

**Comment by tlienart** [comment]
Ok, after playing a bit with it, the proposal seems very sensible. One minor note though:

```julia
df = DataFrames.DataFrame(x=[1,2,3],y=[3,2,1])
df ++ DataFrames.DataFrame(z=[0,1,2])
```

this leads to a concatenation of the two dataframes instead of a larger dataframe (I would have expected it to match `hcat(df, ...)`. This would match

```julia
x = [1 3; 2 2; 3 1]
x ++ [0,1,2]
```

Maybe this:

```julia
++(x::Union{AbstractDataFrame, AbstractVector}...) = hcat(x...)
```

**Comment by CameronBieganek** [comment]
Well, a table isn't really a two-dimensional iterator, it's a one-dimensional iterator of rows, so I would still expect `df1 ++ df2` to perform concatenation of rows.

I'm not sure what is the best way to formalize the definition of `++`. It seems there are a few different possible designs with different tradeoffs.

There's a simple definition for collections that is nice and precise, but overly strict:

```julia
# API #1
++(x::Collection{T}, y::Collection{T})::Collection{T}
```

This definition could be made more general by using type promotion:

```julia
# API #2
++(x::Collection{S}, y::Collection{T})::Collection{promote_type(S, T)}
```

However, that definition still doesn't allow for arbitrary iterators. Would something like the following make sense?

```julia
# API #3
++(itr1, itr2)::Concat{promote_type(eltype(itr1), eltype(itr2))}
```

Note that the third definition of `++` would mean that if you concatenate two vectors, you should get a `Concat` iterator rather than a new vector. And of course that would also be true for data frames. That might feel a little awkward for data frames. I would imagine that most users would expect `df1 ++ df2` to return a regular old data frame rather than a `Concat` iterator.

EDIT:

I suppose `API #3` would probably be better as something like this:

```julia
API #4

itr3 = itr1 ++ itr2
eltype(itr3) = promote_type(eltype(itr1), eltype(itr2)) 
```

That way if `df1 ++ df2` returns a new data frame, it will still be in accordance with API #4.

EDIT \# 2:

Although `API #4` doesn't explicitly specify whether `itr3` is indexable or not. If `itr3` is not guaranteed to be indexable, then `API #4` is essentially my proposal on Discourse for `Iterators.cat`. The current API for `++` allocates and returns new indexable objects.

I don't know if I'm expressing this clearly, but it sort of feels like there are two slightly different concepts: `cat` and `Iterators.cat`. The former corresponds to `API #1` or `API #2` and is the traditional understanding of concatenating vectors or strings, and the latter corresponds to `API #4` (with the specification that `itr3` is iterable but not necessarily indexable). Basically, one version is the lazy version and one version is the eager version.
